src/Maestro/Maestro.Common/AppCredentials/CachedInteractiveBrowserCredential.cs

```
@@ -0,0 +1,83 @@
+        {
+            Directory.CreateDirectory(authRecordDir);
+        }
+
+        if (File.Exists(_authRecordPath))
+        {
+            try
+            {
+                // Fetch existing authentication record to not prompt the user for consent
+                options.AuthenticationRecord = GetAuthenticationRecord();
+                _isCached = true;
+            }
+            catch
+            {
+                // We failed to read the authentication record, we should delete the invalid file
+                File.Delete(_authRecordPath);
+            }
+        }
+
+        _credential = new InteractiveBrowserCredential(options);
+    }
+
+    public override AccessToken GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken)
+    {
+        if (!_isCached)
+        {
+            CacheAuthenticationRecord(requestContext);
+            _isCached = true;
+        }
+
+        return _credential.GetToken(requestContext, cancellationToken);
+    }
+
+    public override ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
+    {
+        if (!_isCached)
+        {
+            CacheAuthenticationRecord(requestContext);
+            _isCached = true;
+        }
+
+        return _credential.GetTokenAsync(requestContext, cancellationToken);
+    }
+
+    private AuthenticationRecord GetAuthenticationRecord()
+    {
+        using var authRecordReadStream = new FileStream(_authRecordPath, FileMode.Open, FileAccess.Read);
+        return AuthenticationRecord.Deserialize(authRecordReadStream);
+    }
+
+    private void CacheAuthenticationRecord(TokenRequestContext requestContext)
+    {
+        // Prompt the user for consent and save the resulting authentication record on disk
+        var authRecord = _credential.Authenticate(requestContext);
+
+        using var authRecordStream = new FileStream(_authRecordPath, FileMode.Create, FileAccess.Write);
```

Should we put some try catch? And behave differently when it is not possible to write AuthRecord to the file system? It still is possible to continue right? So, I am not sure if we should fail hard here in case of file system error.

---

Hmm, if we don't error out here we would be trying to create this file every time Darc is ran.

The assumption here is that since the user is going through interactive creds they would want this cache to be present as to not go through the login process on each command execution. i.e. highlighting this issue immediately and having them resolve it makes sense from my perspective.

But you are right that from Darc's perspective nothing changes if this fails, the command will still run just fine